tests/deploy/ui/test_ui.py

```diff
@@ -0,0 +1,172 @@
+import socket
+import subprocess
+import sys
+import time
+
+import httpx
+import panel as pn
+import pytest
+from playwright.sync_api import Page, expect
+
+from ragna._utils import timeout_after
+from ragna.deploy import Config
+from tests.deploy.utils import TestAssistant
+
+
+def get_available_port():
+    with socket.socket() as s:
+        s.bind(("", 0))
+        return s.getsockname()[1]
+
+
+@pytest.fixture
+def api_port():
+    return get_available_port()
+
+
+@pytest.fixture
+def ui_port():
+    return get_available_port()
+
+
+@pytest.fixture(scope="session")
+def headed_mode(pytestconfig):
+    return pytestconfig.getoption("headed") or False
```

Comment:
Good point. This is something playwright uses already and I was thinking I'd pass it on to ragna in the `open_browser` option but think it would be confusing to have multiple browsers opening during tests. I'll remove it.